Fix crash in linkification reaction, also add linkification to inline markdown parsing (Resolves #2074)

matthew-carroll opened the pull request:
Fix crash in linkification reaction, also add linkification to inline markdown parsing (Resolves #2074)

Repro: Type `[google](www.google.com)`.

Result: Crash due to the (non-Markdown) linkification reaction.

This PR does two things:

1. Don't crash in the non-Markdown linkification reaction.
2. Linkify the Markdown syntax.

Previously, I didn't include Markdown linkification in the upstream Markdown parser reaction. That's because the link syntax is a lot more complicated to parse character by character.

In this PR, I carved out the link syntax as a special case within the parser reaction. In that carve-out I use a `RegExp` to find the link. So now that parser also parses link syntax.

miguelcmedeiros commented:
@matthew-carroll, tested the change and found a problem with the selection after the conversion:

https://github.com/superlistapp/super_editor/assets/31278849/274c4ad4-b056-49e8-96fb-996b08b7eac0

In this case I typed the letter "d" which triggered the conversion, but the selection remains at the end of the link.

matthew-carroll commented:
@miguelcmedeiros I changed the reaction to only linkify when a space is inserted after the link. Please try again.

angelosilvestre reviewed (approved):
LGTM

miguelcmedeiros commented:
@matthew-carroll the selection is still off by -1 when using a space. When the conversion happens, the caret stays at the end of the link and there's a space on the right side.

Also, I noticed that if I copy/paste `[google](www.google.com)`, it throws:
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FormatException: Missing end `]` to match `[` in host (at character 9)
https://[google](www.google.com)
        ^

#0      _Uri._fail (dart:core/uri.dart:1737:5)
#1      _Uri._makeHost (dart:core/uri.dart:2068:9)
#2      new _Uri.notSimple (dart:core/uri.dart:1612:14)
#3      Uri.parse (dart:core/uri.dart:1138:17)
#4      parseLink (package:super_editor/src/default_editor/default_document_editor_reactions.dart:862:13)
#5      PasteEditorCommand._findUrlSpansInText (package:super_editor/src/default_editor/common_editor_operations.dart:2397:21)
#6      PasteEditorCommand._inferAttributionsForLinesOfPastedText (package:super_editor/src/default_editor/common_editor_operations.dart:2370:11)
#7      PasteEditorCommand.execute (package:super_editor/src/default_editor/common_editor_operations.dart:2279:29)
#8      _DocumentEditorCommandExecutor.executeCommand (package:super_editor/src/core/editor.dart:252:15)
#9      Editor._executeCommand (package:super_editor/src/core/editor.dart:195:22)
#10     Editor.execute (package:super_editor/src/core/editor.dart:148:30)
#11     CommonEditorOperations._paste (package:super_editor/src/default_editor/common_editor_operations.dart:2232:12)
```

matthew-carroll commented:
@miguelcmedeiros I corrected the caret placement when linkifying Markdown while you type - I added that condition to the associated test.

I prevented crashing in the standard non-Markdown pasting behavior. I also found that we weren't positioning the caret at the end of pasted text, so I added a selection change there, too. I updated the paste test to also check selection placement.

Please let me know if you find any other weird behaviors.

miguelcmedeiros reviewed (approved):
LGTM

matthew-carroll commented:
@angelosilvestre  do you wanna take another review pass on this?

angelosilvestre commented on super_editor_markdown/lib/src/markdown_inline_upstream_plugin.dart:
```diff
@@ -295,6 +304,66 @@ class _UpstreamInlineMarkdownParser {
+      return null;
+    }
+
+    final markdownLinkRegex = RegExp(r'\[([\w\s\d]+)]\(((?:|https?://)[\w\d./?=#]+)\)');
+    final matches = markdownLinkRegex.allMatches(attributedText.text);
+    if (matches.isEmpty) {
+      // Didn't find any links.
+      return null;
+    }
```
Would it make sense to run the regex only on the portion after the latest space character? For example, if the paragraph text is `Parse the following link: [mylink](https://www.google.com) `.

We could extract `[mylink](https://www.google.com)` based on the spaces before and after it, them we just try to parse that substring. That way we could avoid running the regex over the whole paragraph. We would also make sure that the parsed link is immediately upstream.

matthew-carroll replied:
I thought about that, but I don't actually have a good rule to extract the link. For example, what if the link starts like this " [this is my link](...)"?

angelosilvestre replied:
Understood. 

Maybe we could walk backwards, when we find an empty space:

* If we didn't find any `]` while walking backwards, extract the substring.
* If we found a `]` while walking backwards, keep walking backwards until we find a matching `[`.

Do you think this it would worth the effort?

matthew-carroll replied:
Probably not worth it at this point. I think we can adjust the behavior later if this ends up being too costly. This check should only be running on at most a single paragraph at a time, and only when the user just typed a space.

angelosilvestre replied:
Ok then.

matthew-carroll replied:
Please let me know if you have anything else for the review.

angelosilvestre reviewed (approved):
LGTM
